#!/usr/bin/env python3

import argparse
import logging
import os
import resource
import sys
import traceback

from pyremotenode.schedule import Scheduler
from pyremotenode.utils import Configuration, setup_logging

if __name__ == '__main__':
    a = argparse.ArgumentParser()
    a.add_argument("config", help="Configuration to use for running remote node",
                   type=Configuration.check_file)
    # TODO: Naming of service as a parameter...
    a.add_argument("--start-when-fail", "-s", help="Start even if initial monitor checks fail",
                   action="store_true", default=False)
    a.add_argument("--pidfile", "-p", help="PID file to manage for service operations",
                   default=os.path.join(os.sep, "var", "run", "{0}.pid".format(__name__)))
    a.add_argument("--no-daemon", "-n", help="Do not daemon", default=False, action="store_true")
    args = a.parse_args()

    if not args.no_daemon:
        try:
            pid = os.fork()
            if pid > 0:
                sys.exit(0)
        except OSError as e:
            print("Fork failed: {} ({})".format(e.errno, e.strerror), file=sys.stderr)
            sys.exit(1)

        os.chdir("/")
        os.setsid()
        os.umask(0)

        maxfd = resource.getrlimit(resource.RLIMIT_NOFILE)[1]
        if maxfd == resource.RLIM_INFINITY:
            maxfd = 1024

        # Iterate through and close all file descriptors.
        for fd in range(0, maxfd):
            try:
                os.close(fd)
            except OSError:      # ERROR, fd wasn't open to begin with (ignored)
                pass

        if hasattr(os, "devnull"):
            REDIRECT_TO = os.devnull
        else:
            REDIRECT_TO = "/dev/null"
        os.open(REDIRECT_TO, os.O_RDWR)
        os.dup2(0, 1)
        os.dup2(0, 2)

        # NOTE: Normally to be extra careful we can double fork, but it's not really
        # necessary given we know how this will be instantiated

    cfg = Configuration(args.config).config

    setup_logging("PyRemoteNode",
                  level=getattr(logging, cfg['general']['log_level']),
                  syslog=True)

    try:
        m = Scheduler(cfg,
                      start_when_fail=args.start_when_fail,
                      pid_file=args.pidfile)
        m.run()
    except:
        # Last opportunity to log errors
        logging.error(traceback.format_exc())
